File: From_Last_Year/Theory Lessons Examples/binary_search.cpp
/*"""
binary-SEARCH(A, x)
	return binary-SEARCH(A, x, l, |A|)


binary-SEARCH(A, x, l, r)
	if ell > r then
		return -1
		
	else if A[ell] == x then
		return ell
	else
		return binary-SEARCH(A, x, ell+1, r)
*/

#include <iostream>
#include <vector>

using namespace std;

int binary_search (vector <int> A, int x , int ell, int r);

int binary_search(vector <int> A, int x){
    return binary_search(A, x , 0, A.size()-1);
}

int binary_search (vector <int> A, int x , int ell, int r){
    if (ell > r) return -1;
	
	else{
        int m = (ell+r) / 2;

        if(x < A[m]) return binary_search(A, x, ell, m-1);
        else if(x > A[m]) return binary_search(A, x, m+1, r);
        else return m;
    }
}

	
int main(){

    vector <int> L = {1,2,3,4,5,6,7,8,9,10};

    cout << binary_search(L, 4) << endl;

}

File: From_Last_Year/Theory Lessons Examples/binary_search.py
"""
BINARY-SEARCH(A, x)
	return BINARY-SEARCH(A, x, l, |A|)


BINARY-SEARCH(A, x, l, r)
	if ell > r then
		return -1
		
	else
		m = (ell + r) / 2
			
		else if x < A[m] then
			return BINARY-SEARCH(A, x, ell, m-1)
		
		else if x > A[m] then
			return BINARY-SEARCH(A, x, m+1, r)
		
		else
			return m
"""

def binary_search(A, x):
	return binary_search_rec(A, x , 0, len(A)-1)
	
def binary_search_rec(A, x , ell, r):
	if ell > r:
		return -1
	
	else:
		m = (ell + r) // 2
		
		if x < A[m]:
			return binary_search_rec(A, x, ell, m-1)
		
		if x > A[m]:
			return binary_search_rec(A, x, m+1, r)
	
		else:
			return m
	
L = list(range(1,11))

print(binary_search(L, 4))
